### ISSUE TITLE
How to disable all Printer Function when load a web page with printer button (ex: Gmail) by using <webview>?

### ISSUE THREAD

**Issue by SimonWu29** [electron/electron#13166]
* Electron version: v1.8.4
* Operating system: macOS v10.13.5

From online document, I find call webContents.print() and its silent is true. It will use the default printer device and no pop up dialog.

But I would like to disable all Printer Function when load a web page with printer button (ex: Gmail).
No pop up dialog and No run printer function. 

Please kindly help tell which Listeners could be hook to disable all Printer Function?
Many thanks.

**Comment by MarshallOfSound** [comment]
@SimonWu29 You can set up a preload that runs `window.print = () => {}`, that will stop websites triggering print

**Comment by MarshallOfSound** [comment]
GitHub issues are for feature requests and bug reports, questions about using Electron or code assistance requests should be directed to the [community](https://github.com/electron/electron#community) or to the [Slack Channel](http://atom-slack.herokuapp.com/).

**Comment by sassanh** [comment]
@MarshallOfSound it won't work in iframes, so I guess we can consider it as a feature request (to disable print entirely even in nested iframes without any hacky solution and rather using a clean API)

**Comment by CodeSmith32** [comment]
As @sassanh suggests, and as I agree, this issue should be re-opened since assigning window.print to a noop doesn't prevent printing from nested iframes, and also doesn't provide proper securing from potential pages that would try to print against the user's will.

I would think there should be a hook for preventing / catching print action similar to the 'new-window' event on webContents objects, which effectively allows disabling window.open() in render window and nested iframes.

**Comment by borovinskiy** [comment]
The PDF plugin also cannot be blocked by window.print=()=>{} .

**Comment by ayenter** [comment]
Please re-open this. There is no way to disable print for everywhere.

**Comment by borovinskiy** [comment]
My [browser for kiosks](https://kioskbrowser.elibsystem.ru/) blocking printing as @MarshallOfSound recommended.

@CodeSmith32, тo block printing in frames, you can scan all frames and inject the print blocking code into them too.

Code can be like this (not tested):

```
webContents.mainFrame.framesInSubtree.filter((frame) => {
  frame.executeJavaScript("window.print=()=>{}")
})
```

**Comment by WeijianL1** [comment]
frame.executeJavaScript would trigger `Blocked script execution in 'about:srcdoc' because the document's frame is sandboxed and the 'allow-scripts' permission is not set.` Have you ran into this @borovinskiy ?

**Comment by borovinskiy** [comment]
@WeijianL1,

I am think, but not tested this: 
1) save sandbox attribute of frame
2) add into frame attribute sandbox="allow-scripts" from parent frame
2) inject JS
3) restore sandbox attribute from step 1

**Comment by WeijianL1** [comment]
@borovinskiy 
Thanks for the quick response!
 
I read about this 'allow-scripts' attribute. But that's typically in <iframe> attribute, but for my Electron app it doesn't have any iframe element. The frames might be from the BrowserWindow. Do you know how to add 'allow-scirpt' to the BrowserWindow frames?

**Comment by CodeSmith32** [comment]
> My [browser for kiosks](https://kioskbrowser.elibsystem.ru/) blocking printing as @MarshallOfSound recommended.
> 
> @CodeSmith32, тo block printing in frames, you can scan all frames and inject the print blocking code into them too.
> 
> Code can be like this (not tested):
> 
> ```
> webContents.mainFrame.framesInSubtree.filter((frame) => {
>   frame.executeJavaScript("window.print=()=>{}")
> })
> ```

@borovinskiy, I've tested the code below. When would your code have to run in order to prevent the following code from recreating a working print function, even after `window.print = () => {};` in the top frame of the renderer?

```
function print() {
    var frame = document.createElement('iframe');
    frame.style.cssText = 'width:0; height:0; border:none; position:fixed; z-index:-1000';
    document.body.appendChild(frame);
    frame.contentWindow.print();
    document.body.removeChild(frame);
}
```

That is, how can we inject code immediately after a frame has started loading (synchronously even), but before anything has occurred in it (i.e., in order to modify the JS environment before any JS has run). Is this possible?

**Comment by borovinskiy** [comment]
> > My [browser for kiosks](https://kioskbrowser.elibsystem.ru/) blocking printing as @MarshallOfSound recommended.
> > @CodeSmith32, тo block printing in frames, you can scan all frames and inject the print blocking code into them too.
> > Code can be like this (not tested):
> > ```
> > webContents.mainFrame.framesInSubtree.filter((frame) => {
> >   frame.executeJavaScript("window.print=()=>{}")
> > })
> > ```
> 
> @borovinskiy, I've tested the code below. When would your code have to run in order to prevent the following code from recreating a working print function, even after `window.print = () => {};` in the top frame of the renderer?
> 
> ```
> function print() {
>     var frame = document.createElement('iframe');
>     frame.style.cssText = 'width:0; height:0; border:none; position:fixed; z-index:-1000';
>     document.body.appendChild(frame);
>     frame.contentWindow.print();
>     document.body.removeChild(frame);
> }
> ```
> 
> That is, how can we inject code immediately after a frame has started loading (synchronously even), but before anything has occurred in it (i.e., in order to modify the JS environment before any JS has run). Is this possible?

Thank you for your question. I do not know any methods.

MutationObserver and overwriting document.createElement cannot help in this case.

**Comment by CodeSmith32** [comment]
That's fine, @borovinskiy, which means there really is no currently known solution or work-around, in which case, I *re-assert* that this issue **should be re-opened.**

**Comment by borovinskiy** [comment]
@andreasdj, i am tested 'frame-created' event and in case of @CodeSmith32 it is not work, because when frame created 

`var frame = document.createElement('iframe');`

the frame is not have link to window and code window.print <=> null.print

frame.contentWindow is not null after append to document, but when document appended, it synchronouse print

```
document.body.appendChild(frame);
frame.contentWindow.print();
```

I am not have any thinks for prevent printing by JS in this case.

**Comment by CodeSmith32** [comment]
> I agree that it would be nice to have a built in (less hacky) way to disable the printer functionality altogether.
> 
> If you want to prevent print dialogs from opening inside iframes I still think you can do it by adding a script that runs once a frame is created. Something like:
> 
> ```
> webContents.on('frame-created', (event, details) => {
>     details.frame.executeJavaScript('window.print=()=>{}');
> });  
> ```

@andreasdj, your answer is not hacky at all. In fact, a solution that only selectively prevents printing would be *nice*, but a more ideal solution would be one for generally modifying all [JS *Realms*, if you will,](https://gist.github.com/dherman/7568885#realms) of existing and new frames in the document. Aside from just print, this would allow disabling or overriding many other JS apis (e.g., AudioContext, navigator.geolocation, etc.).

Your solution does in fact run Javascript before the iframe loads. e.g., The following would not work:

```js
var frame = document.createElement('iframe');
frame.src = 'data:text/html;base64,' + btoa('<'+'script>window.print()</'+'script>');
document.body.appendChild(frame);
```

But again, it appears that, since unsandboxed iframes have a fully accessible Realm before loading, accessing `print` out of the `contentWindow` of a new iframe works synchronously. One solution could be to enforce all iframes to be sandboxed. But more ideally, we need a solution for injecting JS synchronously between the lines, even if it's just an adjustment to the timing and usage of the 'frame-created' event.

```js
// ...
document.body.appendChild(frame);
// inject JS into frame
frame.contentWindow.print();
// ...
```
